File: 프로그래머스/3/42898. 등굣길/등굣길.py
# 프로그래머스 42898 등굣길 LV.3
def solution(m, n, puddles):
    answer = 0
    dp = [[1] * m for _ in range(n)]
    denominator = 1000000007
    
    for y, x in puddles:
        if y == 1:
            for i in range(x - 1, n):
                dp[i][y - 1] = 0
        if x == 1:
            for i in range(y - 1, m):
                dp[x - 1][i] = 0
        dp[x - 1][y - 1] = 0
    
    for x in range(n):
        for y in range(m):
            if 0 <= x - 1 <= n and 0 <= y - 1 <= m and dp[x][y] != 0:
                dp[x][y] = dp[x - 1][y] + dp[x][y - 1]
                
    return dp[-1][-1] % denominator

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;

class Solution {
    int[][] dp;
    
    public int solution(int m, int n, int[][] puddles) {
        dp = new int[n][m];
        int denominator = 1000000007;
        
        for (int[] x : dp) {
            Arrays.fill(x, 1);
        }
        
        for (int[] puddle : puddles) {
            int py = puddle[0] - 1, px = puddle[1] - 1;
            dp[px][py] = 0;
            
            if (py == 0) {
                for (int ii = px; ii < n; ii++) {
                    dp[ii][py] = 0;
                }
            }
            
            if (px == 0) {
                for (int jj = py; jj < m; jj++) {
                    dp[px][jj] = 0;
                }
            }
        }

        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                if (dp[i][j] != 0) {
                    dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % denominator;
                }
            }
        }
        
        return dp[n - 1][m - 1];
    }
}